File: ui/src/server/model-catalog.ts
// WHAT THE PROVIDER ALREADY TOLD US ABOUT EACH MODEL, cached and read back.
//
// THE PROBLEM THIS SOLVES, and it was costing real verdicts. Talaria asked four
// hundred models' worth of catalog from OpenRouter on every endpoint edit,
// parsed it down to `string[]`, and threw the rest away. So the fitness page
// said "nothing advertises a context window" about a model whose catalog entry
// says 1,048,576 tokens, and left `tools`, `json`, `json-strict` and `vision`
// unmeasured — capping every slot at `workable` and asking an admin to buy
// probes — for four facts the provider publishes for free.
//
// It is a CACHE, not a source of truth. `provider-catalog.ts` does the fetching
// and stays the only thing that talks to a provider; this module decides how
// long an answer is good for, where it lives, and what a capability reader may
// conclude from it. A stale entry is served while a refresh is in flight, and a
// failed refresh serves the last good answer — the same posture `openrouterUsPool`
// takes, for the same reason: a catalog blip must not empty the model picker.
//
// ── THE ONE RULE THAT MATTERS ────────────────────────────────────────────────
// A CATALOG MAY SAY YES. IT MAY NEVER SAY NO.
//
// `supported_parameters` listing `tools` is a provider committing to something,
// and Talaria records it as a declared fact. `supported_parameters` NOT listing
// `tools` is not a denial — it is a field a hundred OpenAI-compatible servers
// never populate at all, plus vLLM, Ollama, LM Studio and every self-host that
// answers `/models` with `{id}` and nothing else.
//
// And the asymmetry is not merely cautious, it is load-bearing: `run.ts` step 2
// refuses a run outright on a capability recorded false from a `declared`
// source. Writing "false" from an absent field would take a model the platform
// has never tested, decide it cannot call tools because a self-hosted server did
// not fill in an optional array, and REFUSE the judge on it. Absence stays
// unknown, `capability.ts`'s cardinal rule (UNKNOWN IS NOT FALSE) keeps the
// model running, and a probe is what turns unknown into a no.
import { getSetting, setSetting } from './audit'
import { listEndpoints, type LlmEndpoint } from './agent-defs'
import { capabilityKey, mergeCapabilities, type Capability, type CapabilityFact } from './harness/capability'
import { catalogModels, type CatalogModel } from './provider-catalog'

/** One endpoint's catalog, as of `at`. Keyed by endpoint NAME because that is
 *  what `capabilityKey` and the gateway's routing both speak. */
export interface EndpointCatalog {
  endpoint: string
  at: string
  models: CatalogModel[]
}

export type CatalogStore = Record<string, EndpointCatalog>

const KEY = 'model_catalog'

/** A CATALOG IS NOT VOLATILE, and a day is the honest number. Providers add
 *  models continuously and change a model's advertised window almost never, so
 *  a shorter TTL buys nothing but request volume against four hundred rows. The
 *  admin-facing refresh button exists for the case where it matters. */
export const CATALOG_TTL_MS = 24 * 60 * 60 * 1000

export interface CatalogDeps {
  endpoints: () => Promise<LlmEndpoint[]>
  fetchCatalog: (ep: LlmEndpoint) => Promise<CatalogModel[]>
  read: () => Promise<CatalogStore>
  write: (store: CatalogStore) => Promise<void>
  /** Merged into `model_capabilities` in ONE write, respecting source
   *  precedence. Returns how many facts were actually taken — a catalog claim
   *  that loses to an existing probe result is not a write. */
  record: (batch: Array<{ key: string; facts: Partial<Record<Capability, CapabilityFact>> }>) => Promise<number>
  now: () => number
}

const REAL: CatalogDeps = {
  endpoints: listEndpoints,
  fetchCatalog: catalogModels,
  read: () => getSetting<CatalogStore>(KEY, {}),
  write: (store) => setSetting(KEY, store),
  record: mergeCapabilities,
  now: () => Date.now(),
}

const withDeps = (over?: Partial<CatalogDeps>): CatalogDeps => ({ ...REAL, ...over })

// ── Reading ──────────────────────────────────────────────────────────────────

/** Everything known about one model id, across every endpoint that serves it.
 *
 *  MORE THAN ONE ENTRY IS NORMAL and is why this returns a list rather than a
 *  row: a bare id can land on any endpoint in the pool, and the two can differ
 *  in exactly the way that matters (a quantized local build next to the vendor's
 *  own API). Callers reduce it themselves — `advertisedWindow` takes the
 *  smallest, because a claim has to hold for the worst member. */
export async function catalogEntriesFor(model: string, deps?: Partial<CatalogDeps>): Promise<Array<{ endpoint: string; model: CatalogModel }>> {
  const store = await withDeps(deps).read().catch((): CatalogStore => ({}))
  const out: Array<{ endpoint: string; model: CatalogModel }> = []
  for (const cat of Object.values(store)) {
    const hit = cat.models.find((m) => m.id === model)
    if (hit) out.push({ endpoint: cat.endpoint, model: hit })
  }
  return out
}

/** THE SMALLEST advertised window across the pool, or null when nothing
 *  advertises one. Same rule the probe suite already applied to the endpoint
 *  row; the difference is that this reads the number the provider published
 *  ABOUT THIS MODEL rather than one integer stamped on the endpoint. */
export async function advertisedWindow(model: string, deps?: Partial<CatalogDeps>): Promise<number | null> {
  const windows = (await catalogEntriesFor(model, deps)).map((e) => e.model.contextLength).filter((n): n is number => typeof n === 'number' && n > 0)
  return windows.length ? Math.min(...windows) : null
}

// ── Deriving capabilities ────────────────────────────────────────────────────

/** Which advertised parameter proves which capability. Only entries whose
 *  PRESENCE is a real commitment belong here — see the file header on why
 *  absence proves nothing. */
const PARAM_PROVES: ReadonlyArray<{ param: string; capability: Capability; detail: string }> = [
  { param: 'tools', capability: 'tools', detail: 'the provider catalog advertises tool calling' },
  { param: 'response_format', capability: 'json', detail: 'the provider catalog advertises response_format' },
  { param: 'structured_outputs', capability: 'json-strict', detail: 'the provider catalog advertises structured outputs' },
  { param: 'web_search_options', capability: 'search', detail: 'the provider catalog advertises native web search' },
]

/** WHAT A CATALOG ENTRY PROVES, and nothing more.
 *
 *  Pure, exported, and tested directly, because this function is where the
 *  yes-only rule either holds or quietly stops holding. Every branch below
 *  writes `true` or writes NOTHING. There is no path that writes `false`. */
export function capabilitiesFromCatalog(m: CatalogModel, at: string): Partial<Record<Capability, CapabilityFact>> {
  const facts: Partial<Record<Capability, CapabilityFact>> = {}
  // `source: 'catalog'`, never 'declared'. 'declared' means a HUMAN said so and
  // outranks a provider's own claim — see `SOURCE_RANK` in capability.ts. Writing
  // these as 'declared' would let a nightly refresh overrule an admin.
  const declare = (capability: Capability, detail: string): void => {
    facts[capability] = { value: true, source: 'catalog', at, detail }
  }

  const params = m.supportedParameters
  if (params) {
    for (const rule of PARAM_PROVES) {
      if (params.includes(rule.param)) declare(rule.capability, rule.detail)
    }
    // `tool_choice` without `tools` is not a thing any catalog publishes, but a
    // model that accepts it can certainly be handed tools.
    if (params.includes('tool_choice') && !facts.tools) declare('tools', 'the provider catalog advertises tool_choice')
  }

  // VISION IS THE ONE PLACE ABSENCE IS ALMOST INFORMATIVE and still is not
  // recorded as a no: a catalog that lists `input_modalities: ['text']` is
  // describing the model precisely, but plenty of catalogs list nothing at all,
  // and one shape must not be read as the other. A probe closes it.
  if (m.inputModalities?.includes('image')) declare('vision', 'the provider catalog advertises image input')

  // `long-context` IS DELIBERATELY NOT DECLARED HERE, and the temptation to do
  // it was the whole reason to write this down. `context_length` is the only
  // catalog field that looks like a capability and is not one: it says what the
  // model will ACCEPT, and the probe measures whether the model can still find
  // a fact planted in the middle of that window. Advertising 1M tokens and
  // retrieving from 1M tokens are different claims, and the gap between them is
  // exactly what an admin is trying to find out.
  //
  // The window is not wasted — `advertisedWindow` serves it to the probe, which
  // is what decides how big a haystack to build. That is the honest use of a
  // spec sheet: it sizes the measurement, it does not stand in for it.
  return facts
}

// ── Refreshing ───────────────────────────────────────────────────────────────

export interface RefreshResult {
  endpoint: string
  models: number
  /** Capability facts written across all of this endpoint's models. */
  facts: number
  error: string | null
}

/** Re-fetch one endpoint's catalog, store it, and declare what it proves.
 *
 *  A FAILED FETCH KEEPS THE OLD ENTRY. An endpoint that is briefly unreachable
 *  must not empty the picker or un-declare facts an admin is looking at; the
 *  error is reported and the last good catalog stays. */
export async function refreshEndpointCatalog(ep: LlmEndpoint, deps?: Partial<CatalogDeps>): Promise<RefreshResult> {
  const d = withDeps(deps)
  const at = new Date(d.now()).toISOString()
  let models: CatalogModel[]
  try {
    models = await d.fetchCatalog(ep)
  } catch (err) {
    return { endpoint: ep.name, models: 0, facts: 0, error: err instanceof Error ? err.message : String(err) }
  }

  const store = await d.read().catch((): CatalogStore => ({}))
  await d.write({ ...store, [ep.name]: { endpoint: ep.name, at, models } })

  // Per endpoint:model, exactly as every other capability writer keys it — a
  // fact learned about one endpoint's build is never credited to another's. The
  // whole endpoint goes in ONE merge: four hundred separate read-modify-writes
  // of a row that grows with each one is not a thing to do on a cadence.
  const batch = models
    .map((m) => ({ key: capabilityKey(ep.name, m.id), facts: capabilitiesFromCatalog(m, at) }))
    .filter((e) => Object.keys(e.facts).length > 0)
  const facts = await d.record(batch).catch(() => 0)
  return { endpoint: ep.name, models: models.length, facts, error: null }
}

/** Every endpoint whose catalog is older than the TTL (or absent). Called on a
 *  cadence and by the admin's refresh button; `force` ignores the TTL. */
export async function refreshCatalogs(opts: { force?: boolean } = {}, deps?: Partial<CatalogDeps>): Promise<RefreshResult[]> {
  const d = withDeps(deps)
  const [endpoints, store] = await Promise.all([d.endpoints().catch((): LlmEndpoint[] => []), d.read().catch((): CatalogStore => ({}))])
  const now = d.now()
  const stale = endpoints.filter((ep) => {
    if (opts.force) return true
    const at = Date.parse(store[ep.name]?.at ?? '')
    return !Number.isFinite(at) || now - at > CATALOG_TTL_MS
  })
  // SEQUENTIAL. Four hundred rows per provider is a big response and several
  // providers rate-limit the catalog endpoint harder than they rate-limit
  // completions; a fan-out here buys nothing an admin can perceive.
  const out: RefreshResult[] = []
  for (const ep of stale) out.push(await refreshEndpointCatalog(ep, deps))
  return out
}

File: api/src/model_catalog.rs
// WHAT THE PROVIDER ALREADY TOLD US ABOUT EACH MODEL, cached and read back —
// port of model-catalog.ts (read half + the one refresh the effort backfill
// needs). It is a CACHE, not a source of truth: provider.rs does the fetching
// and stays the only thing that talks to a provider; this module decides how
// long an answer is good for, where it lives, and what a capability reader
// may conclude from it. A stale entry is served while a refresh is in flight,
// and a failed refresh serves the last good answer — a catalog blip must not
// empty the model picker.
//
// ── THE ONE RULE THAT MATTERS ────────────────────────────────────────────────
// A CATALOG MAY SAY YES. IT MAY NEVER SAY NO.
//
// `supported_parameters` listing `tools` is a provider committing to
// something, and Talaria records it as a catalog fact. NOT listing `tools` is
// not a denial — it is a field a hundred OpenAI-compatible servers never
// populate at all, plus vLLM, Ollama, LM Studio and every self-host that
// answers `/models` with `{id}` and nothing else. Absence stays unknown,
// capability.rs's cardinal rule (UNKNOWN IS NOT FALSE) keeps the model
// running, and a probe is what turns unknown into a no.
//
// `advertised_window` and the full `refresh_catalogs` sweep port with the
// probe suite and the admin refresh button (the surfaces that own them); the
// effort ladder's needs stop at what is here.

use crate::capability::{CapabilityFact, capability_key, merge_capabilities};
use crate::gateway::provider::CatalogModel;
use crate::gateway::registry::LlmEndpoint;
use crate::gateway::settings::{get_setting, set_setting};
use sqlx::PgPool;
use std::collections::HashSet;
use std::time::{SystemTime, UNIX_EPOCH};

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

pub const KEY: &str = "model_catalog";

/// A CATALOG IS NOT VOLATILE, and a day is the honest number. Providers add
/// models continuously and change a model's advertised window almost never,
/// so a shorter TTL buys nothing but request volume against four hundred rows.
pub const CATALOG_TTL_MS: u64 = 24 * 60 * 60 * 1000;

// ── Reading ──────────────────────────────────────────────────────────────────

/// Everything known about one model id, across every endpoint that serves it
/// — port of catalogEntriesFor. MORE THAN ONE ENTRY IS NORMAL and is why this
/// returns a list: a bare id can land on any endpoint in the pool, and the
/// two can differ in exactly the way that matters (a quantized local build
/// next to the vendor's own API). Callers reduce it themselves.
pub async fn catalog_entries_for(pg: &PgPool, model: &str) -> Vec<(String, CatalogModel)> {
    entries_in_store(&read_store(pg).await, model)
}

/// The pure core of `catalog_entries_for`, over an already-read store.
///
/// BOTH SPELLINGS, and the endpoint-qualified one is why this exists at all.
/// A catalog is keyed by the id the PROVIDER publishes — OpenRouter says
/// `deepseek/deepseek-v4-flash` — while everything that asks about a model
/// says `openrouter/deepseek/deepseek-v4-flash`, the one spelling the picker
/// offers. The prefix is stripped only for the endpoint whose name it is, so
/// a bare id still searches every catalog (the pooled case) and a qualified
/// id can never match another endpoint's entry.
pub fn entries_in_store(store: &serde_json::Value, model: &str) -> Vec<(String, CatalogModel)> {
    let mut out = Vec::new();
    let Some(obj) = store.as_object() else {
        return out;
    };
    for (name, entry) in obj {
        let models = entry.get("models").and_then(|m| m.as_array());
        let Some(models) = models else { continue };
        let prefix = format!("{name}/");
        let upstream = model.strip_prefix(&prefix).unwrap_or(model);
        for raw in models {
            let Some(m) = CatalogModel::from_json(raw) else {
                continue;
            };
            if m.id == upstream {
                out.push((name.clone(), m));
                break;
            }
        }
    }
    out
}

/// Catalog entries for EXPLICIT endpoint:upstream pairs — the persona path.
/// A fleet persona id is not a catalog id, so it cannot be looked up the way
/// `entries_in_store` looks one up; its agent config names the exact targets
/// (persona.rs), and this answers the catalog rows for those targets and
/// nothing else: prefix-stripping here would be guessing at what the config
/// already knows.
pub async fn catalog_entries_for_targets(
    pg: &PgPool,
    targets: &[crate::persona::ModelTarget],
) -> Vec<(String, CatalogModel)> {
    targets_in_store(&read_store(pg).await, targets)
}

pub fn targets_in_store(
    store: &serde_json::Value,
    targets: &[crate::persona::ModelTarget],
) -> Vec<(String, CatalogModel)> {
    let mut out = Vec::new();
    let obj = store.as_object();
    for t in targets {
        let Some(entry) = obj.and_then(|o| o.get(&t.endpoint)) else {
            continue;
        };
        let Some(models) = entry.get("models").and_then(|m| m.as_array()) else {
            continue;
        };
        for raw in models {
            let Some(m) = CatalogModel::from_json(raw) else {
                continue;
            };
            if m.id == t.model {
                out.push((t.endpoint.clone(), m));
                break;
            }
        }
    }
    out
}

async fn read_store(pg: &PgPool) -> serde_json::Value {
    get_setting(pg, KEY, serde_json::json!({})).await
}

// ── Reasoning effort ─────────────────────────────────────────────────────────

/// The effort ladder's display order, weakest to strongest. Levels a provider
/// coins that this list does not know still show — appended after the known
/// ones, in the provider's own order — because the provider's spelling is the
/// contract and the picker must never rename a level into one the model
/// rejects.
pub const EFFORT_ORDER: [&str; 7] = ["none", "minimal", "low", "medium", "high", "xhigh", "max"];

pub fn ordered_efforts(levels: &[String]) -> Vec<String> {
    let mut out = Vec::new();
    let mut seen: HashSet<String> = HashSet::new();
    for known in EFFORT_ORDER {
        if levels.iter().any(|l| l == known) && seen.insert(known.to_string()) {
            out.push(known.to_string());
        }
    }
    for l in levels {
        if !EFFORT_ORDER.contains(&l.as_str()) && seen.insert(l.clone()) {
            out.push(l.clone());
        }
    }
    out
}

/// THE EFFORT LEVELS A POOL AGREES ON — port of effortLevelsOf.
///
/// Same posture as a minimum window: a bare id (and a persona behind
/// fallbacks) can land on any member of the pool, so a level is only offered
/// when every member that speaks at all accepts it. Members that say nothing
/// do not veto — UNKNOWN IS NOT FALSE — they are skipped, exactly as an
/// endpoint without a published window does not shrink the minimum. An empty
/// answer means "nobody vouched for any level" and the picker stays hidden,
/// which is the honest rendering of an unmeasured model.
pub fn effort_levels_of(entries: &[(String, CatalogModel)]) -> Vec<String> {
    let lists: Vec<Vec<String>> = entries
        .iter()
        .filter_map(|(_, m)| m.efforts.as_ref().filter(|l| !l.is_empty()).cloned())
        .collect();
    let Some((first, rest)) = lists.split_first() else {
        return Vec::new();
    };
    let mut shared = first.clone();
    for l in rest {
        shared.retain(|x| l.contains(x));
    }
    ordered_efforts(&shared)
}

// ── Deriving capabilities ────────────────────────────────────────────────────

/// Which advertised parameter proves which capability. Only entries whose
/// PRESENCE is a real commitment belong here — see the file header on why
/// absence proves nothing.
const PARAM_PROVES: [(&str, &str, &str); 4] = [
    (
        "tools",
        "tools",
        "the provider catalog advertises tool calling",
    ),
    (
        "response_format",
        "json",
        "the provider catalog advertises response_format",
    ),
    (
        "structured_outputs",
        "json-strict",
        "the provider catalog advertises structured outputs",
    ),
    (
        "web_search_options",
        "search",
        "the provider catalog advertises native web search",
    ),
];

/// WHAT A CATALOG ENTRY PROVES, and nothing more — port of
/// capabilitiesFromCatalog. Pure, exported, and tested directly, because this
/// function is where the yes-only rule either holds or quietly stops holding.
/// Every branch below pushes a `true` or pushes NOTHING. There is no path
/// that writes `false`.
pub fn capabilities_from_catalog(m: &CatalogModel, at: &str) -> Vec<(String, CapabilityFact)> {
    let mut facts: Vec<(String, CapabilityFact)> = Vec::new();
    let declare = |facts: &mut Vec<(String, CapabilityFact)>, cap: &str, detail: &str| {
        // `source: "catalog"`, never "declared". "declared" means a HUMAN said
        // so and outranks a provider's own claim — see capability.rs. Writing
        // these as "declared" would let a nightly refresh overrule an admin.
        facts.push((
            cap.to_string(),
            CapabilityFact {
                value: true,
                source: "catalog".to_string(),
                at: at.to_string(),
                detail: Some(detail.to_string()),
                score: None,
            },
        ));
    };
    if let Some(params) = &m.supported_parameters {
        for (param, cap, detail) in PARAM_PROVES {
            if params.iter().any(|p| p == param) {
                declare(&mut facts, cap, detail);
            }
        }
        // `tool_choice` without `tools` is not a thing any catalog publishes,
        // but a model that accepts it can certainly be handed tools.
        if params.iter().any(|p| p == "tool_choice") && !facts.iter().any(|(cap, _)| cap == "tools")
        {
            declare(
                &mut facts,
                "tools",
                "the provider catalog advertises tool_choice",
            );
        }
    }
    // VISION IS THE ONE PLACE ABSENCE IS ALMOST INFORMATIVE and still is not
    // recorded as a no: a catalog that lists `input_modalities: ['text']` is
    // describing the model precisely, but plenty of catalogs list nothing at
    // all, and one shape must not be read as the other. A probe closes it.
    if m.input_modalities
        .as_ref()
        .is_some_and(|mods| mods.iter().any(|x| x == "image"))
    {
        declare(
            &mut facts,
            "vision",
            "the provider catalog advertises image input",
        );
    }
    // `long-context` IS DELIBERATELY NOT DECLARED HERE. `context_length` is
    // the only catalog field that looks like a capability and is not one: it
    // says what the model will ACCEPT, and the probe measures whether the
    // model can still find a fact planted in the middle of that window.
    // Advertising 1M tokens and retrieving from 1M tokens are different
    // claims, and the gap between them is exactly what an admin is trying to
    // find out. The window is not wasted — `advertised_window` (probe suite,
    // later batch) serves it to the probe, which sizes the measurement.
    facts
}

// ── Refreshing ───────────────────────────────────────────────────────────────

pub struct RefreshResult {
    pub endpoint: String,
    pub models: usize,
    /// Capability facts written across all of this endpoint's models.
    pub facts: usize,
    pub error: Option<String>,
}

/// Re-fetch one endpoint's catalog, store it, and declare what it proves —
/// port of refreshEndpointCatalog.
///
/// A FAILED FETCH KEEPS THE OLD ENTRY. An endpoint that is briefly
/// unreachable must not empty the picker or un-declare facts an admin is
/// looking at; the error is reported and the last good catalog stays.
pub async fn refresh_endpoint_catalog(
    state: &crate::state::AppState,
    ep: &LlmEndpoint,
) -> RefreshResult {
    // `new Date().toISOString()` — millisecond precision, Z.
    let at = crate::agent_auth::epoch_ms_to_iso(now_ms() as i64);
    let models = match crate::gateway::provider::catalog_models(state, ep).await {
        Ok(m) => m,
        Err(err) => {
            return RefreshResult {
                endpoint: ep.name.clone(),
                models: 0,
                facts: 0,
                error: Some(err),
            };
        }
    };

    let mut store = read_store(&state.pg).await;
    let entry = serde_json::json!({
        "endpoint": ep.name,
        "at": at,
        "models": models.iter().map(CatalogModel::to_json).collect::<Vec<_>>(),
    });
    if let Some(obj) = store.as_object_mut() {
        // Preserve-order insert: an existing key keeps its position (the TS
        // spread's behavior), a new one appends.
        obj.insert(ep.name.clone(), entry);
    }
    if set_setting(&state.pg, KEY, &store).await.is_err() {
        // TS throws here and the caller decides; the fetch succeeded but the
        // last good catalog stays until the next attempt.
        return RefreshResult {
            endpoint: ep.name.clone(),
            models: 0,
            facts: 0,
            error: Some("could not store the refreshed catalog".to_string()),
        };
    }

    // Per endpoint:model, exactly as every other capability writer keys it — a
    // fact learned about one endpoint's build is never credited to another's.
    // The whole endpoint goes in ONE merge: four hundred separate
    // read-modify-writes of a row that grows with each one is not a thing to
    // do on a cadence.
    let batch: Vec<(String, Vec<(String, CapabilityFact)>)> = models
        .iter()
        .map(|m| {
            (
                capability_key(&ep.name, &m.id),
                capabilities_from_catalog(m, &at),
            )
        })
        .filter(|(_, facts)| !facts.is_empty())
        .collect();
    let facts = merge_capabilities(&state.pg, &batch).await.unwrap_or(0);
    RefreshResult {
        endpoint: ep.name.clone(),
        models: models.len(),
        facts,
        error: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn model(id: &str, efforts: Option<Vec<&str>>) -> serde_json::Value {
        let mut o = serde_json::Map::new();
        o.insert("id".into(), json!(id));
        o.insert("name".into(), json!(null));
        o.insert("contextLength".into(), json!(null));
        o.insert("inputModalities".into(), json!(null));
        o.insert("supportedParameters".into(), json!(null));
        o.insert(
            "efforts".into(),
            match efforts {
                None => json!(null),
                Some(l) => json!(l),
            },
        );
        o.insert("pricing".into(), json!(null));
        serde_json::Value::Object(o)
    }

    fn store(cats: &[(&str, &[serde_json::Value])]) -> serde_json::Value {
        let mut o = serde_json::Map::new();
        for (name, models) in cats {
            o.insert(
                (*name).to_string(),
                json!({"endpoint": name, "at": "2026-01-01T00:00:00.000Z", "models": models}),
            );
        }
        serde_json::Value::Object(o)
    }

    #[test]
    fn qualified_ids_match_only_their_own_endpoint() {
        let s = store(&[
            ("openrouter", &[model("deepseek/v4", None)]),
            ("local", &[model("deepseek/v4", None)]),
        ]);
        // Qualified: only openrouter's entry, prefix stripped.
        let hits = entries_in_store(&s, "openrouter/deepseek/v4");
        assert_eq!(hits.len(), 1);
        assert_eq!(hits[0].0, "openrouter");
        assert_eq!(hits[0].1.id, "deepseek/v4");
        // Bare: every catalog that serves it.
        assert_eq!(entries_in_store(&s, "deepseek/v4").len(), 2);
        // A qualified id must not match another endpoint's entry, even when
        // the vendor prefix looks like one: "local/deepseek/v4" as an
        // OPENROUTGER catalog id is matched only by that literal id.
        assert_eq!(entries_in_store(&s, "local/deepseek/v4").len(), 1);
        assert!(entries_in_store(&s, "nowhere/x").is_empty());
    }

    #[test]
    fn the_pool_intersects_and_silent_members_do_not_veto() {
        let e = |ep: &str, efforts: Option<Vec<&str>>| {
            (
                ep.to_string(),
                CatalogModel::from_json(&model("m", efforts)).unwrap(),
            )
        };
        // Both speak: intersection.
        let levels = effort_levels_of(&[
            e("a", Some(vec!["low", "high"])),
            e("b", Some(vec!["high", "max"])),
        ]);
        assert_eq!(levels, vec!["high".to_string()]);
        // A silent member does not veto.
        let levels = effort_levels_of(&[e("a", Some(vec!["low"])), e("b", None)]);
        assert_eq!(levels, vec!["low".to_string()]);
        // Nobody speaks: empty.
        assert!(effort_levels_of(&[e("a", None), e("b", None)]).is_empty());
        assert!(effort_levels_of(&[]).is_empty());
        // Unknown levels append in provider order after the known ones.
        let levels = effort_levels_of(&[e("a", Some(vec!["turbo", "high", "eco"]))]);
        assert_eq!(
            levels,
            vec!["high".to_string(), "turbo".to_string(), "eco".to_string()]
        );
    }

    #[test]
    fn a_catalog_may_say_yes_never_no() {
        let with_params = |params: Option<Vec<&str>>, mods: Option<Vec<&str>>| CatalogModel {
            id: "m".into(),
            name: None,
            context_length: Some(1_048_576.0),
            input_modalities: mods.map(|m| m.iter().map(|s| s.to_string()).collect()),
            supported_parameters: params.map(|p| p.iter().map(|s| s.to_string()).collect()),
            efforts: None,
            pricing: None,
        };
        let at = "2026-01-01T00:00:00.000Z";
        // Silence writes NOTHING — not one false.
        let facts = capabilities_from_catalog(&with_params(None, None), at);
        assert!(facts.is_empty());
        let facts = capabilities_from_catalog(&with_params(Some(vec![]), Some(vec!["text"])), at);
        assert!(facts.is_empty());
        // Each advertised param proves exactly its capability.
        let facts = capabilities_from_catalog(
            &with_params(
                Some(vec![
                    "tools",
                    "response_format",
                    "structured_outputs",
                    "web_search_options",
                ]),
                Some(vec!["text", "image"]),
            ),
            at,
        );
        let caps: Vec<&str> = facts.iter().map(|(c, _)| c.as_str()).collect();
        assert_eq!(
            caps,
            vec!["tools", "json", "json-strict", "search", "vision"]
        );
        assert!(facts.iter().all(|(_, f)| f.value && f.source == "catalog"));
        // tool_choice alone proves tools; long-context is never declared.
        let facts = capabilities_from_catalog(&with_params(Some(vec!["tool_choice"]), None), at);
        let caps: Vec<&str> = facts.iter().map(|(c, _)| c.as_str()).collect();
        assert_eq!(caps, vec!["tools"]);
        // tools already present: tool_choice does not double-declare.
        let facts =
            capabilities_from_catalog(&with_params(Some(vec!["tools", "tool_choice"]), None), at);
        assert_eq!(facts.len(), 1);
    }
}
